galley/pkg/crd/validation/validation.go

```diff
@@ -0,0 +1,139 @@
+
+
+
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package validation
+
+import (
+"fmt"
+"net"
+"net/http"
+"github.com/prometheus/client_golang/prometheus/promhttp"
+
+"istio.io/istio/galley/cmd/shared"
+"istio.io/istio/mixer/adapter"
+"istio.io/istio/mixer/pkg/config"
+"istio.io/istio/mixer/pkg/config/store"
+runtimeConfig "istio.io/istio/mixer/pkg/runtime/config"
+"istio.io/istio/mixer/pkg/template"
+generatedTmplRepo "istio.io/istio/mixer/template"
+"istio.io/istio/pilot/pkg/model"
+"istio.io/istio/pkg/cmd"
+"istio.io/istio/pkg/kube"
+"istio.io/istio/pkg/log"
+"istio.io/istio/pkg/version"
+)
+
+// createMixerValidator creates a mixer backend validator.
+// TODO(https://github.com/istio/istio/issues/4887) - refactor mixer
+// config validation to remove galley dependency on mixer internal
+// packages.
+func createMixerValidator() (store.BackendValidator, error) {
+	info := generatedTmplRepo.SupportedTmplInfo
+	templates := make(map[string]*template.Info, len(info))
+	for k := range info {
+		t := info[k]
+		templates[k] = &t
+	}
+	adapters := config.AdapterInfoMap(adapter.Inventory(), template.NewRepository(info).SupportsTemplate)
+	return store.NewValidator(nil, runtimeConfig.KindMap(adapters, templates)), nil
+}
+
+
+const (
+	metricsPath = "/metrics"
+	versionPath = "/version"
+)
+
+func startSelfMonitoring(stop <-chan struct{}, port uint) {
```

Comment:
We should be able to share the promhttp handler between the validation and server paths. Consider moving `startSelfMonitoring` to galley/pkg/server/monitoring.go?

Reply:
Now there is no metrics available in server, so  keep the logics for validation only  move it once server have added metrics part?

Reply:
We'll be adding metrics shortly. Since you're already moving this code around we might as well put it the common place now :)

Reply:
Done~ Move to galley/pkg/server/galleymonitor.go